### PR TITLE
Traffic log admin

### PR DESCRIPTION
**Pull request by lund5000**
🚨 Requires `traffic_log_admin` role in chirpradiodev to use these features

Trello: https://trello.com/c/nKp9Ay7i

In this PR:
- List spots and their copy
- Add/edit/delete spots and the schedule for when they should be read
- Add/edit/delete the different copy to be read
- Set the start and end dates for copy in bulk
- Delete copy in bulk

Out-of-scope for this PR:
- The traffic log interface for DJs
- Recreating the report on traffic log activity
- Tagging copy and filtering on those tags
- Separating the concepts of active and deleted on spots
- Showing expired copy
  - currently the DJDB treats expired and deleted as the same thing, so that'll take a little more untangling

### CODE REVIEW

File: app/index.yaml [hunk 1]
```diff
@@ -64,4 +64,9 @@ indexes:
...
   - name: dow
   - name: hour
-  - name: created
\ No newline at end of file
+  - name: created
+
+- kind: Spot
+  properties:
+  - name: active
+  - name: deleted
```
Comment (lund5000) [inline app/index.yaml:72]: Datastore indexes allow you to run queries with particular combinations of property filters.

File: client/src/router/index.js [hunk 2]
```diff
@@ -190,14 +191,7 @@ const routes = [
...
   },
-  {
-    path: "/traffic-log",
-    component: () => import("../views/traffic-log/TrafficLog.vue"),
-  },
-  {
-    path: "/traffic-log/spots",
-    component: () => import("../views/traffic-log/SpotsView.vue"),
-  },
+  ...trafficLogRoutes,
```
Comment (lund5000) [inline client/src/router/index.js:194]: This PR starts to break up the Vue client by module. In this case the new route definitions are co-located with the other /traffic-log code then imported and spread into the `routes` array (rather than added to an ever-expanding file with all the route definitions).

File: client/src/router/index.js [hunk 3]
```diff
@@ -212,12 +206,20 @@ const router = createRouter({
   },
 });
 
-router.beforeEach((to, from, next) => {
```
Comment (lund5000) [inline client/src/router/index.js:215]: Use of `next` has been deprecated in more recent versions of the router, so I fix that here while also adding a guard that prevents an unauthorized user from directly navigating to a route.

https://router.vuejs.org/guide/advanced/navigation-guards.html#Optional-third-argument-next

File: client/src/router/index.js [hunk 4]
```diff
@@ -212,12 +206,20 @@ const router = createRouter({
...
+  }
+
+  if (to.meta.requiresAuth && !authStore.isAuthorized(to.meta.requiresAuth)) {
+    return false;
+  }
+});
+
+router.afterEach((to) => {
+  if (to.meta.title) {
+    document.title = `${to.meta.title} – NextUp`;
```
Comment (lund5000) [inline client/src/router/index.js:222]: Sometimes the title update is dynamic and is best handled in the view once data is loaded. But when it's simple and  static, we can set the string on the route itself.

File: client/src/services/api.service.js [hunk 5]
```diff
@@ -35,6 +35,8 @@ async function getAndHandleError(getter) {
   return response.data;
 }
 
+export { instance as api };
```
Comment (lund5000) [inline client/src/services/api.service.js:38]: Big change in this file. Similar to the changes above where I start slimming down the router and co-locating code into a /traffic-log module, I do the same thing here with the API-related code. Exporting the `instance` allows the stores to make API calls directly.

File: client/src/services/api.service.js [hunk 6]
```diff
@@ -177,54 +176,4 @@ export default {
     });
     return await getAndHandleError(getter);
   },
-
```
Comment (lund5000) [inline client/src/services/api.service.js:180]: All the removed API calls below show up later and get updated in the /traffic-log/store. This keeps related code closer together and removes an extra jump between the store this service when trying to follow the flow of data.

File: client/src/store/modules/spots.js
Comment (lund5000) [inline client/src/store/modules/spots.js:1]: This was a remnant of the previous way of defining stores before switching to the now-recommended `pinia` library.

File: client/src/stores/auth.js [hunk 7]
```diff
@@ -8,10 +8,14 @@ export const useAuthStore = defineStore("auth", {
...
-      //   roles: [],
-      // },
+      /**
+       * template: {
+       *   users: [],
+       *   roles: [],
+       **/
+      "traffic-log": {
+        roles: ["traffic_log_admin"],
+      },
```
Comment (lund5000) [inline client/src/stores/auth.js:18]: Super basic feature flagging

File: client/src/traffic-log/functions.js
Comment (lund5000) [inline client/src/traffic-log/functions.js:1]: These helper functions were needed in more than one file, so I put them here to try and keep things DRY

File: client/src/traffic-log/store.js [hunk 8]
```diff
@@ -0,0 +1,134 @@
+import { defineStore } from "pinia";
+import { api } from "../services/api.service";
```
Comment (lund5000) [inline client/src/traffic-log/store.js:2]: Imports the configured Axios instance so it can be used directly in the methods below. Keeping the API paths abstracted away into that service didn't seem nearly as useful as keeping the logic closer together and easier to read.

File: client/vite.config.js [hunk 9]
```diff
@@ -5,6 +5,9 @@ const path = require("path");
 // https://vitejs.dev/config/
 export default defineConfig({
   plugins: [vue()],  
+  define: {
+    __VUE_PROD_HYDRATION_MISMATCH_DETAILS__: 'false',
```
Comment (lund5000) [inline client/vite.config.js:9]: I was getting an angry-looking error by not having this explicitly defined.

https://vuejs.org/api/compile-time-flags.html#VUE_PROD_HYDRATATION_MISMATCH_DETAILS

File: client/vite.config.js [hunk 10]
```diff
@@ -44,9 +47,11 @@ export default defineConfig({
             "./src/views/crates/CratesView.vue",
             "./src/views/crates/CrateView.vue",            
           ], 
-          "traffic-log": [
-            "./src/views/traffic-log/SpotsView.vue",
-            "./src/views/traffic-log/TrafficLog.vue",
+          "traffic-log": [
```
Comment (lund5000) [inline client/vite.config.js:50]: This is where the JS can be configured to be compiled into separate chunks that are smaller and loaded only when needed.

File: app/services/spot.service.js [hunk 11]
```diff
@@ -99,14 +99,33 @@ async function updateSpot(id, data) {
 async function deleteSpot(id) {
   await updateSpot(id, {
     active: false,
+    deleted: true,
```
Comment (lund5000) [inline app/services/spot.service.js:102]: The DJDB and database currently treats spots where `active === false` as deleted. Shawn wants to be able to temporarily flag a spot as inactive without losing all the schedule configuration she's already done for it. This PR adds a new `deleted` property to the spot, but it'll take future PRs (and in this case all the DJs beginning to use the traffic log in NextUp) to split up the two concepts.

File: app/services/spot.service.js [hunk 12]
```diff
@@ -129,6 +150,7 @@ async function updateCopy(id, data) {
 async function deleteCopy(id) {
   return await updateCopy(id, {
     expire_on: new Date(),
+    deleted: true,
```
Comment (lund5000) [inline app/services/spot.service.js:153]: Similar to spots, the DJDB treats any expired copy as deleted. This starts to lay the groundwork for separating those two ideas so a traffic log admin can still find and rescheduled expired copy.

**Review by bmcdonn2 — APPROVED** [review]
Full disclosure: this is my first exposure to Vue. Having said that, the changes aren't too hard to follow. The shift to a module-based structure does improve readability. Overall, the code appears to meet the requirements.

I ran the code locally and all functionality covered by this PR appears to be covered without any noticeable bugs.
